File: apps/mobile_flutter/lib/features/progress/domain/progress_repository.dart
import '../../../core/api/api_client.dart';
import '../../language_selection/domain/language_pack.dart';
import '../../profiles/domain/learner_profile.dart';
import 'learned_word.dart';
import 'progress_event.dart';
import 'progress_summary.dart';

class ProgressRepository {
  const ProgressRepository(this.apiClient);

  final ApiClient apiClient;

  Future<ProgressSummary> fetchSummary({
    required LearnerProfile profile,
    required LanguagePack language,
  }) async {
    // Scoped to the selected language -- see BV-PROGRESS-SCOPE-001. An
    // unscoped/global completed_activities count could exceed a smaller
    // language's own catalog size and corrupt the lesson-journey
    // next_index/state calc, making every activity in that language look
    // already completed.
    final json = await apiClient.getMap(
      '/profiles/${profile.id}/progress?language_code=${Uri.encodeComponent(language.code)}',
    );
    return ProgressSummary.fromJson(
      profileName: profile.displayName,
      languageName: language.name,
      json: json,
    );
  }

  Future<List<LearnedWord>> fetchLearnedWords({
    required LearnerProfile profile,
    required LanguagePack language,
  }) async {
    final json = await apiClient.getMap(
      '/profiles/${profile.id}/learned-words?language_code=${language.code}',
    );
    final words = json['words'] as List<dynamic>? ?? const [];
    return [
      for (final word in words)
        LearnedWord.fromJson(word as Map<String, dynamic>),
    ];
  }

  Future<void> recordActivityCompleted({
    required LearnerProfile profile,
    required LanguagePack language,
    required String activityId,
    required int clientSequence,
  }) async {
    final event = await buildActivityCompletedEvent(
      profile: profile,
      language: language,
      activityId: activityId,
      clientSequence: clientSequence,
    );
    await uploadEvent(event);
  }

  ProgressEvent buildActivityCompletedEvent({
    required LearnerProfile profile,
    required LanguagePack language,
    required String activityId,
    required int clientSequence,
  }) {
    return ProgressEvent.activityCompleted(
      profileId: profile.id,
      languageCode: language.code,
      activityId: activityId,
      clientSequence: clientSequence,
    );
  }

  Future<void> uploadEvent(ProgressEvent event) async {
    await apiClient.postMap(
      '/progress/events',
      event.toJson(),
    );
  }
}

File: apps/web_pwa/src/features/progress/progressRepository.ts
import type { ApiClient } from '../../core/api/apiClient';
import type { LearnerProfile } from '../profiles/types';
import type { LanguagePack } from '../languages/types';
import {
  buildActivityCompletedEvent,
  learnedWordFromJson,
  progressEventToJson,
  progressSummaryFromJson,
  type LearnedWord,
  type ProgressEvent,
  type ProgressSummary,
} from './types';

// Mirrors apps/mobile_flutter/lib/features/progress/domain/progress_repository.dart.
export class ProgressRepository {
  private readonly apiClient: ApiClient;

  constructor(apiClient: ApiClient) {
    this.apiClient = apiClient;
  }

  async fetchSummary({
    profile,
    language,
  }: {
    profile: LearnerProfile;
    language: LanguagePack;
  }): Promise<ProgressSummary> {
    // Scoped to the selected language so "Completed" reflects that
    // language's own catalog, not a cross-language total (see
    // BV-PROGRESS-SCOPE-001 -- a global count could exceed a smaller
    // catalog's size and made every activity look already completed).
    const json = await this.apiClient.getMap<Record<string, unknown>>(
      `/profiles/${profile.id}/progress?language_code=${encodeURIComponent(language.code)}`,
    );
    return progressSummaryFromJson(profile.displayName, language.name, json);
  }

  async fetchLearnedWords({
    profile,
    language,
  }: {
    profile: LearnerProfile;
    language: LanguagePack;
  }): Promise<LearnedWord[]> {
    const json = await this.apiClient.getMap<Record<string, unknown>>(
      `/profiles/${profile.id}/learned-words?language_code=${encodeURIComponent(language.code)}`,
    );
    const words = (json.words as Record<string, unknown>[] | undefined) ?? [];
    return words.map(learnedWordFromJson);
  }

  buildActivityCompletedEvent(args: {
    profile: LearnerProfile;
    language: LanguagePack;
    activityId: string;
    clientSequence: number;
  }): ProgressEvent {
    return buildActivityCompletedEvent(args);
  }

  async uploadEvent(event: ProgressEvent): Promise<void> {
    await this.apiClient.postMap('/progress/events', progressEventToJson(event));
  }
}
